### PR TITLE
Feature/update

### PR DESCRIPTION
**Pull request by stefbenoist**
- Allow to run update operation with Yorc
- Update resources and modifiers
- Use ProxySelector to yorc HTTP connection (to handle http;nonProxyHosts setting)
- Fix issue on operation artifact long/short notations use in Velocity templates
- Add configuration parameters to the driver configuration in order to support TLS connection with Yorc

### CODE REVIEW

**Comment by adidanes** [comment]
The branch feature/configure-tls-connection merged in this branch.

**Review by adidanes — APPROVED** [review]
Tested with Welcome sample in a secured installation with GCP as target infrastrucure.

File: src/main/java/alien4cloud/paas/yorc/context/rest/TemplateManager.java [hunk 1]
```diff
@@ -79,11 +95,38 @@ public void configure() throws PluginConfigurationException {
...
+            } catch (NoSuchAlgorithmException | KeyStoreException | IOException | CertificateException | InvalidKeySpecException e) {
+                e.printStackTrace();
+                throw new PluginConfigurationException("Failed to create keystore", e);
+            }
+            try {
+                context = SSLContext.getInstance("TLS");
+                context.init(getKeyManagers(keystore), getTrustManagers(keystore), null);
+            } catch (NoSuchAlgorithmException | KeyStoreException | UnrecoverableKeyException
+                    | KeyManagementException e) {
+                e.printStackTrace();
```
Comment (xdegenne) [inline src/main/java/alien4cloud/paas/yorc/context/rest/TemplateManager.java]: avoid such e.printStackTrace(), please use logger

File: src/main/java/alien4cloud/paas/yorc/context/rest/TemplateManager.java [hunk 2]
```diff
@@ -79,11 +95,38 @@ public void configure() throws PluginConfigurationException {
...
-        SSLContext context = SSLContexts.createSystemDefault();
+        if(!isSSLConfigProvided()) {
+            context = SSLContexts.createSystemDefault();
+        } else {
+            // In order to configure the SSLContext we need to create a keystore
+            KeyStore keystore;
+            try {
+                keystore = createKeystore();
+            } catch (NoSuchAlgorithmException | KeyStoreException | IOException | CertificateException | InvalidKeySpecException e) {
+                e.printStackTrace();
```
Comment (xdegenne) [inline src/main/java/alien4cloud/paas/yorc/context/rest/TemplateManager.java]: avoid such e.printStackTrace(), please use logger

**Review by xdegenne — CHANGES_REQUESTED** [review]

File: src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmBuilder.java [hunk 3]
```diff
@@ -65,6 +65,49 @@ private void configure(StateMachineTransitionConfigurer<FsmStates, FsmEvents> tr
...
+			.source(FsmStates.UPDATED).target(FsmStates.UNDEPLOYMENT_IN_PROGRESS)
+			.event(FsmEvents.UNDEPLOYMENT_STARTED)
+			.action(actions.undeploy())
+// Allow undeploy from UPDATE_FAILED state
+		.and()
+			.withExternal()
+			.source(FsmStates.UPDATE_FAILED).target(FsmStates.UNDEPLOYMENT_IN_PROGRESS)
+			.event(FsmEvents.UNDEPLOYMENT_STARTED)
+			.action(actions.undeploy())
+
```
Comment (SebastienLaurent) [inline src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmBuilder.java:110]: Missing transitions on EVICTION event

File: src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmMapper.java [hunk 4]
```diff
@@ -83,7 +88,12 @@ public static FsmStates fromYorcToFsmState(String status) throws Exception {
             case "DEPLOYMENT_IN_PROGRESS":
                 return FsmStates.DEPLOYMENT_IN_PROGRESS;
             case "DEPLOYMENT_FAILED":
-                return FsmStates.FAILED;
```
Comment (SebastienLaurent) [inline src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmMapper.java]: Why is the "DePLOYMENT_FAILED" state now mapped to UPDATE_IN_PROGRESS FSM State?

File: src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmBuilder.java (hunk 3 above)
Comment (adidanes) [reply src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmBuilder.java:110]: Can you please give more details ?

File: src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmMapper.java (hunk 4 above)
Comment (adidanes) [reply src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmMapper.java]: I think its an oversite... 
I can fix it if if you think its necessary. Stephane its in vacation until early september, so we cannot wait for him to discuss.

Comment (SebastienLaurent) [reply src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmMapper.java]: Yes i think it needs a fix

File: src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmBuilder.java (hunk 3 above)
Comment (SebastienLaurent) [reply src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmBuilder.java:110]: We need to add Eviction transitions by inserting this:

```suggestion
		.and()
			.withExternal()
			.source(FsmStates.UPDATE_FAILED).target(FsmStates.UNDEPLOYED)
			.event(FsmEvents.EVICTION)
			.action(actions.cleanup())
		.and()
			.withExternal()
			.source(FsmStates.UPDATED).target(FsmStates.UNDEPLOYED)
			.event(FsmEvents.EVICTION)
			.action(actions.cleanup())
		.and()
			.withExternal()
			.source(FsmStates.UPDATE_IN_PROGRESS).target(FsmStates.UNDEPLOYED)
			.event(FsmEvents.EVICTION)
			.action(actions.cleanup())			
```

When a deployment is alive (in a state other than UNDEPLOYED) in a4c but does not exist any more in yorc, the eviction thread use theses transitions to set the deployment as UNDEPLOYED

Comment (adidanes) [reply src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmBuilder.java:110]: OK. Fixed.

File: src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmMapper.java (hunk 4 above)
Comment (adidanes) [reply src/main/java/alien4cloud/paas/yorc/context/service/fsm/FsmMapper.java]: Fixed.

**Review by SebastienLaurent — APPROVED** [review]
Looks ok for me

**Pull request merged**